### PR TITLE
Fix node deletion

### PR DESCRIPTION
**Pull request by patrickdbakke**
# Summary
Fixes node addition `.add()`, _which was **very subtly** buggy in very edge-casey ways that probably never happen in the real_ world. 
Fixes node deletion `.delete()` which was buggy in more obvious ways. 

# Changes
* rescopes a variable in `.add()`: `var item` --> `let item`
* fixes the `index === key.length` deletion case
* stops errantly converting non-leaf nodes into broken "leaf" nodes when deleting.
* lints whitespace
* lotsa unit tests

# Todo:
Bump new version, release.

### CODE REVIEW

File: lib/node.js [hunk 1]
```diff
@@ -175,13 +175,15 @@ Node.prototype.delete = function (key, distinct) {
...
     // If the index matches the key length, we are at the node to be deleted.
     // Remove values associated with the key, where the `distinct` value also matches.
-    this.values = this.values.filter((item) => (item.key !== key && item.distinct !== distinct));
+    this.values = this.values.filter(
+      (item) => item.key !== key && item.distinct !== distinct
+    );
 
     // If the node becomes a leaf node, update its status.
-    this.leaf = this.values.length > 0;
+    // this.leaf = this.values.length > 0;
```
Comment (patrickdbakke) [inline lib/node.js]: To confirm - I don't think we actually care about changing leaf to non-leaf when deleting. It'll _**maybe**_ save a recursive loop when querying, but isn't strictly necessary.

File: lib/node.js [hunk 2]
```diff
@@ -175,13 +175,15 @@ Node.prototype.delete = function (key, distinct) {
 };
 
 Node.prototype._deleteRecursive = function (key, distinct, index) {
-  if (index === key.length || this.leaf) {
+  if (this.leaf) {
```
Comment (patrickdbakke) [inline lib/node.js:178]: Non-trivial change here.

File: lib/node.js [hunk 3]
```diff
@@ -199,11 +201,16 @@ Node.prototype._deleteRecursive = function (key, distinct, index) {
...
+  this.children[chr] = this.children[chr]._deleteRecursive(
+    key,
+    distinct,
+    index + 1
+  );
 
   // If the child node has no values or children after deletion, remove it.
   if (!this.children[chr]) {
     delete this.children[chr];
+    this.values.splice(index, 1);
```
Comment (patrickdbakke) [inline lib/node.js]: Non-trivial change here.

File: test/trie_test.js [hunk 4]
```diff
@@ -13,9 +13,9 @@ describe("Trie", function () {
 
     t.add(item);
 
-    assert.equal(1, t.prefixSearch("o")[0]);
-    assert.equal(1, t.prefixSearch("on")[0]);
-    assert.equal(1, t.prefixSearch("one")[0]);
```
Comment (patrickdbakke) [inline test/trie_test.js:18]: Reordered all asserts to correctly use `expected` vs `actual` order.

File: lib/node.js [hunk 5]
```diff
@@ -192,18 +191,33 @@ Node.prototype._deleteRecursive = function (key, distinct, index) {
     return this;
   }
 
+  if (index === key.length) {
```
Comment (patrickdbakke) [inline lib/node.js:194]: Non-trivial change here.

File: lib/node.js [hunk 6]
```diff
@@ -45,7 +45,7 @@ Node.prototype.add = function (item, index, maxWidth) {
     this.values = [];
     this.leaf = false;
     for (var i = 0; i < oldValues.length; i++) {
-      var item = oldValues[i],
+      let item = oldValues[i],
```
Comment (patrickdbakke) [inline lib/node.js:48]: Non-trivial change here. [See this comment for some storytelling](https://superhuman.slack.com/archives/C05N2TSFVDX/p1695362866382519).

File: lib/node.js [hunk 7]
```diff
@@ -175,13 +175,12 @@ Node.prototype.delete = function (key, distinct) {
...
-  if (index === key.length || this.leaf) {
+  if (this.leaf) {
     // If the index matches the key length, we are at the node to be deleted.
     // Remove values associated with the key, where the `distinct` value also matches.
-    this.values = this.values.filter((item) => (item.key !== key && item.distinct !== distinct));
-
-    // If the node becomes a leaf node, update its status.
-    this.leaf = this.values.length > 0;
+    this.values = this.values.filter(
+      (item) => !(item.key === key && item.distinct === distinct)
```
Comment (patrickdbakke) [inline lib/node.js:182]: Non-trivial change here (fixing `distinct` comparison)

File: lib/node.js [hunk 8]
```diff
@@ -192,18 +191,33 @@ Node.prototype._deleteRecursive = function (key, distinct, index) {
...
+    key,
+    distinct,
+    index + 1
+  );
 
   // If the child node has no values or children after deletion, remove it.
   if (!this.children[chr]) {
     delete this.children[chr];
+    const charIdx = this.values.indexOf(chr);
+    if (charIdx >= 0) {
```
Comment (patrickdbakke) [inline lib/node.js:218]: Non-trivial change here.

Comment (danbits) [reply lib/node.js:218]: Thanks for pointing out the specific changes to focus on. It's really helpful!

File: lib/node.js [hunk 9]
```diff
@@ -192,18 +191,33 @@ Node.prototype._deleteRecursive = function (key, distinct, index) {
     return this;
   }
 
+  if (index === key.length) {
+    if (this.children[undefined]) {
+      this.children[undefined]._deleteRecursive(key, distinct, index + 1);
+    }
```
Comment (danbits) [inline lib/node.js:197]: This is to resolve the initial edge case you mentioned with `undefined` keys, right?

**Review by danbits — APPROVED** [review]
Thanks for digging deep into this trie and resolving lots of edge cases!

File: lib/node.js (hunk 9 above)
Comment (patrickdbakke) [reply lib/node.js:197]: Yes exactly. 
the `undefined` key is used for non-leaf nodes to identify that [this item is exactly this key]


Visually: 
```
// skipping "children" and "values" for brevity
{
   a: {
       undefined: {}  we have an item(s) with key *exactly* "a" and nothing longer
       a: {...} // we have an item with prefix "aa", but don't yet know if its shorter or not (need to look deeper)
   }
   b: {
      b: {
         undefined: {}  we have an item(s) with key *exactly* "bb" and not longer
         a: {...} // we have an item with prefix "bba", but don't yet know if its shorter or not (need to look deeper)
      }
   }
}
```

Comment (danbits) [reply lib/node.js:197]: Thanks for the example, that makes perfect sense!

File: lib/node.js [hunk 10]
```diff
@@ -175,13 +175,12 @@ Node.prototype.delete = function (key, distinct) {
...
 
 Node.prototype._deleteRecursive = function (key, distinct, index) {
-  if (index === key.length || this.leaf) {
+  if (this.leaf) {
     // If the index matches the key length, we are at the node to be deleted.
     // Remove values associated with the key, where the `distinct` value also matches.
-    this.values = this.values.filter((item) => (item.key !== key && item.distinct !== distinct));
-
-    // If the node becomes a leaf node, update its status.
-    this.leaf = this.values.length > 0;
```
Comment (haydenlinder) [inline lib/node.js:184]: Does the removal of this line cover this?
> stops errantly converting non-leaf nodes into broken "leaf" nodes when deleting.

Comment (patrickdbakke) [reply lib/node.js:184]: Yes exactly!

**Review by haydenlinder — APPROVED** [review]

**Pull request merged**